# Вебинар 6. Двухуровневые модели рекомендаций


### Зачем 2 уровня?
- Классические модели классификации (lightgbm) зачастую работают лучше, чем рекоммендательные модели (als, lightfm)
- Данных много, предсказаний много (# items * # users) --> с таким объемом lightgbm не справляется
- Но рекомендательные модели справляются!

Отбираем top-N (200) *кандидатов* с помощью простой модели (als) --> переранжируем их сложной моделью (lightgbm)
и выберем top-k (10).

---

### Как отбирать кандидатов?

Вариантов множество. Тут нам поможет *MainRecommender*. Пока в нем реализованы далеко не все возможные способы генерации кандидатов

- Генерируем топ-k кандидатов
- Качество кандидатов измеряем через **recall@k**
- recall@k показывает какую долю из купленных товаров мы смогли выявить (рекомендовать) нашей моделью

----

# Практическая часть

# Import libs

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Для работы с матрицами
from scipy.sparse import csr_matrix

# Матричная факторизация
from implicit import als

# Модель второго уровня
from lightgbm import LGBMClassifier

import os, sys
module_path = os.path.abspath(os.path.join(os.pardir))
if module_path not in sys.path:
    sys.path.append(module_path)

# Написанные нами функции
from metrics import precision_at_k, recall_at_k
from utils import prefilter_items
from recommenders import MainRecommender

## Read data

In [3]:
data = pd.read_csv('../data/retail_train.csv')
item_features = pd.read_csv('../data/product.csv')
user_features = pd.read_csv('../data/hh_demographic.csv')

# Process features dataset

In [4]:
ITEM_COL = 'item_id'
USER_COL = 'user_id'

In [5]:
# column processing
item_features.columns = [col.lower() for col in item_features.columns]
user_features.columns = [col.lower() for col in user_features.columns]

item_features.rename(columns={'product_id': ITEM_COL}, inplace=True)
user_features.rename(columns={'household_key': USER_COL }, inplace=True)

# Split dataset for train, eval, test

In [6]:
# Важна схема обучения и валидации!
# -- давние покупки -- | -- 6 недель -- | -- 3 недель -- 
# подобрать размер 2-ого датасета (6 недель) --> learning curve (зависимость метрики recall@k от размера датасета)


VAL_MATCHER_WEEKS = 6
VAL_RANKER_WEEKS = 3

In [7]:
# берем данные для тренировки matching модели
data_train_matcher = data[data['week_no'] < data['week_no'].max() - (VAL_MATCHER_WEEKS + VAL_RANKER_WEEKS)]

# берем данные для валидации matching модели
data_val_matcher = data[(data['week_no'] >= data['week_no'].max() - (VAL_MATCHER_WEEKS + VAL_RANKER_WEEKS)) &
                      (data['week_no'] < data['week_no'].max() - (VAL_RANKER_WEEKS))]


# берем данные для тренировки ranking модели
data_train_ranker = data_val_matcher.copy()  # Для наглядности. Далее мы добавим изменения, и они будут отличаться

# берем данные для теста ranking, matching модели
data_val_ranker = data[data['week_no'] >= data['week_no'].max() - VAL_RANKER_WEEKS]

In [8]:
def print_stats_data(df_data, name_df):
    print(name_df)
    print(f"Shape: {df_data.shape} Users: {df_data[USER_COL].nunique()} Items: {df_data[ITEM_COL].nunique()}")

In [9]:
print_stats_data(data_train_matcher,'train_matcher')
print_stats_data(data_val_matcher,'val_matcher')
print_stats_data(data_train_ranker,'train_ranker')
print_stats_data(data_val_ranker,'val_ranker')

train_matcher
Shape: (2108779, 12) Users: 2498 Items: 83685
val_matcher
Shape: (169711, 12) Users: 2154 Items: 27649
train_ranker
Shape: (169711, 12) Users: 2154 Items: 27649
val_ranker
Shape: (118314, 12) Users: 2042 Items: 24329


In [10]:
# выше видим разброс по пользователям и товарам

In [11]:
data_train_matcher.head(2)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0         -0.6        1631        1          0.0                0.0  
1          0.0        1631        1          0.0                0.0

# Prefilter items

In [12]:
n_items_before = data_train_matcher['item_id'].nunique()

data_train_matcher = prefilter_items(data_train_matcher, item_features=item_features, take_n_popular=5000)

n_items_after = data_train_matcher['item_id'].nunique()
print('Decreased # items from {} to {}'.format(n_items_before, n_items_after))

C:\Users\Al\RecSystems\Lesson6\utils.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['price'] = data['sales_value'] / (np.maximum(data['quantity'], 1))


Decreased # items from 83685 to 5001


# Make cold-start to warm-start

In [13]:
# ищем общих пользователей
common_users = data_train_matcher.user_id.values

data_val_matcher = data_val_matcher[data_val_matcher.user_id.isin(common_users)]
data_train_ranker = data_train_ranker[data_train_ranker.user_id.isin(common_users)]
data_val_ranker = data_val_ranker[data_val_ranker.user_id.isin(common_users)]

print_stats_data(data_train_matcher,'train_matcher')
print_stats_data(data_val_matcher,'val_matcher')
print_stats_data(data_train_ranker,'train_ranker')
print_stats_data(data_val_ranker,'val_ranker')

train_matcher
Shape: (861404, 13) Users: 2495 Items: 5001
val_matcher
Shape: (169615, 12) Users: 2151 Items: 27644
train_ranker
Shape: (169615, 12) Users: 2151 Items: 27644
val_ranker
Shape: (118282, 12) Users: 2040 Items: 24325


In [14]:
# Теперь warm-start по пользователям

# Init/train recommender

In [15]:
recommender = MainRecommender(data_train_matcher)

### Варианты, как получить кандидатов

Можно потом все эти варианты соединить в один

(!) Если модель рекомендует < N товаров, то рекомендации дополняются топ-популярными товарами до N

In [16]:
# Берем тестового юзера 2375

In [17]:
recommender.get_als_recommendations(2375, N=5)

[899624, 1044078, 871756, 844179, 1106523]

In [18]:
recommender.get_own_recommendations(2375, N=5)

[948640, 918046, 847962, 907099, 873980]

In [19]:
recommender.get_similar_items_recommendation(2375, N=5)

[1046545, 1044078, 937292, 842125, 15778319]

In [20]:
# recommender.get_similar_users_recommendation(2375, N=5)

# Eval recall of matching

### Измеряем recall@k

Это будет в ДЗ: 

A) Попробуйте различные варианты генерации кандидатов. Какие из них дают наибольший recall@k ?
- Пока пробуем отобрать 50 кандидатов (k=50)
- Качество измеряем на data_val_matcher: следующие 6 недель после трейна

Дают ли own recommendtions + top-popular лучший recall?  

B)* Как зависит recall@k от k? Постройте для одной схемы генерации кандидатов эту зависимость для k = {20, 50, 100, 200, 500}  
C)* Исходя из прошлого вопроса, как вы думаете, какое значение k является наиболее разумным?


In [21]:
ACTUAL_COL = 'actual'

In [22]:
result_eval_matcher = data_val_matcher.groupby(USER_COL)[ITEM_COL].unique().reset_index()
result_eval_matcher.columns=[USER_COL, ACTUAL_COL]
result_eval_matcher.head(2)

user_id                                             actual
0        1  [853529, 865456, 867607, 872137, 874905, 87524...
1        2  [15830248, 838136, 839656, 861272, 866211, 870...

In [23]:
# N = Neighbors
N_PREDICT = 50 

In [24]:
%%time
# для понятности расписано все в строчку, без функций, ваша задача уметь оборачивать все это в функции
result_eval_matcher['own_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))
result_eval_matcher['sim_item_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_similar_items_recommendation(x, N=50))
result_eval_matcher['als_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_als_recommendations(x, N=50))

Wall time: 29.9 s


In [25]:
%%time
#result_eval_matcher['sim_user_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_similar_users_recommendation(x, N=50))

Wall time: 0 ns


### Пример оборачивания

In [26]:
# # сырой и простой пример как можно обернуть в функцию
def evalRecall_webinar(df_result, target_col_name, recommend_model, result_col_name, num_cand):
    result_col_name = 'result'
    df_result[result_col_name] = df_result[target_col_name].apply(lambda x: recommend_model(x, N=50))
    return df_result.apply(lambda row: recall_at_k(row[result_col_name], row[ACTUAL_COL], k=N_PREDICT), axis=1).mean()

In [27]:
# # сырой и простой пример как можно обернуть в функцию
def evalRecall(df_result, col_name, num_cand, recommend_model, target_col_name):
    result_col_name = 'result'
    df_result[result_col_name] = df_result[target_col_name].apply(lambda x: recommend_model(x, N=num_cand))
    return df_result.apply(lambda row: recall_at_k(row[result_col_name], row[ACTUAL_COL], k=num_cand), axis=1).mean()

In [28]:
import itertools
num_cand_list = [20, 50, 100, 200, 500, 1000]
var_list = (['own_rec', recommender.get_own_recommendations],
                 ['sim_item_rec', recommender.get_similar_items_recommendation],
                 ['als_rec', recommender.get_als_recommendations]
           )

params_list = list(itertools.product(var_list, num_cand_list))
params_list

[(['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  20),
 (['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  50),
 (['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  100),
 (['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  200),
 (['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  500),
 (['own_rec',
   <bound method MainRecommender.get_own_recommendations of <recommenders.MainRecommender object at 0x0000029A805A3C10>>],
  1000),
 (['sim_item_rec',
   <bound method MainRecommender.get_similar_items_recommendation of <recommenders.MainRecommender object at 0x0

In [29]:
val_rec_list = []

In [30]:
%%time
for i in range(len(params_list)):
    #print(params_list[i][0][0])
    #print(params_list[i][0][1])
    #print(params_list[i][1])      
    val_rec = evalRecall(result_eval_matcher, params_list[i][0][0], params_list[i][1], params_list[i][0][1], USER_COL)
    #print(val_rec)
    val_rec_list.append(val_rec)

Wall time: 4min 20s


val_rec_list

In [31]:
columns_params = ['name_var', 'num_cand']
df_params = pd.DataFrame(params_list, columns = columns_params)
#df_params
columns_pres = ['recall']
df_recall = pd.DataFrame(val_rec_list, columns = columns_pres)

all_columns = ['name_var', 'num_cand', 'recall']
df_rezult = pd.concat([df_params, df_recall], ignore_index=True, axis = 1)
df_rezult.columns = all_columns
df_rezult.sort_values('recall', ascending=False, inplace=True)

df_rezult

name_var  num_cand    recall
5   [own_rec, <bound method MainRecommender.get_ow...      1000  0.216791
17  [als_rec, <bound method MainRecommender.get_al...      1000  0.195856
4   [own_rec, <bound method MainRecommender.get_ow...       500  0.182053
11  [sim_item_rec, <bound method MainRecommender.g...      1000  0.181654
16  [als_rec, <bound method MainRecommender.get_al...       500  0.146028
10  [sim_item_rec, <bound method MainRecommender.g...       500  0.136257
3   [own_rec, <bound method MainRecommender.get_ow...       200  0.135373
15  [als_rec, <bound method MainRecommender.get_al...       200  0.097144
2   [own_rec, <bound method MainRecommender.get_ow...       100  0.096045
9   [sim_item_rec, <bound method MainRecommender.g...       200  0.085937
14  [als_rec, <bound method MainRecommender.get_al...       100  0.069350
1   [own_rec, <bound method MainRecommender.get_ow...        50  0.065257
8   [sim_item_rec, <bound method MainRecommender.g...       100  0.054427
13  [als_rec, <bound method MainRecommender.get_al...        50  0.047365
0   [own_rec, <bound method MainRecommender.get_ow...        20  0.039284
7   [sim_item_rec, <bound method MainRecommender.g...        50  0.033598
12  [als_rec, <bound method MainRecommender.get_al...        20  0.029186
6   [sim_item_rec, <bound method MainRecommender.g...        20  0.017598

In [32]:
#лучший результат показа own_rec. Исследуем более подробно.

In [33]:
var_list_own_rec = [recommender.get_own_recommendations]

num_cand_list_own_rec = [20, 50, 100, 200, 500, 600, 700, 800, 900, 1000]

In [34]:
val_rec_list_own_rec = []

In [35]:
%%time
for i in range(len(num_cand_list_own_rec)):
    #print(params_list[i][0][0])
    #print(params_list[i][0][1])
    #print(params_list[i][1])      
    val_rec = evalRecall(result_eval_matcher, "own_rec", num_cand_list_own_rec[i], recommender.get_own_recommendations, USER_COL)
    #print(val_rec)
    val_rec_list_own_rec.append(val_rec)

Wall time: 1min 23s


In [36]:
columns_params = ['num_cand']
df_params = pd.DataFrame(num_cand_list_own_rec, columns = columns_params)

columns_rec = ['recall']
df_recall = pd.DataFrame(val_rec_list_own_rec, columns = columns_rec)

all_columns = ['num_cand', 'recall']
df_rezult = pd.concat([df_params, df_recall], ignore_index=True, axis = 1)
df_rezult.columns = all_columns
df_rezult.sort_values('recall', ascending=False, inplace=True)

df_rezult

num_cand    recall
9      1000  0.216791
8       900  0.210944
7       800  0.203604
6       700  0.197028
5       600  0.189985
4       500  0.182053
3       200  0.135373
2       100  0.096045
1        50  0.065257
0        20  0.039284

In [37]:
# рост значений recall замедляется при количестве ~ 500.

In [38]:
# # сырой и простой пример как можно обернуть в функцию
def evalRecall(df_result, target_col_name, recommend_model):
    result_col_name = 'result'
    df_result[result_col_name] = df_result[target_col_name].apply(lambda x: recommend_model(x, N=50))
    return df_result.apply(lambda row: recall_at_k(row[result_col_name], row[ACTUAL_COL], k=N_PREDICT), axis=1).mean()

In [39]:
def calc_recall(df_data, top_k):
    for col_name in df_data.columns[2:]:
        yield col_name, df_data.apply(lambda row: recall_at_k(row[col_name], row[ACTUAL_COL], k=top_k), axis=1).mean()

In [40]:
def calc_precision(df_data, top_k):
    for col_name in df_data.columns[2:]:
        yield col_name, df_data.apply(lambda row: precision_at_k(row[col_name], row[ACTUAL_COL], k=top_k), axis=1).mean()

### Recall@50 of matching

In [41]:
TOPK_RECALL = 50

In [42]:
sorted(calc_recall(result_eval_matcher, TOPK_RECALL), key=lambda x: x[1],reverse=True)

[('own_rec', 0.06525657038145175),
 ('result', 0.06525657038145175),
 ('als_rec', 0.04736545034000402),
 ('sim_item_rec', 0.033598469816240296)]

### Precision@5 of matching

In [43]:
TOPK_PRECISION = 5

In [44]:
sorted(calc_precision(result_eval_matcher, TOPK_PRECISION), key=lambda x: x[1],reverse=True)

[('own_rec', 0.17712691771268974),
 ('result', 0.17712691771268974),
 ('als_rec', 0.11650395165039425),
 ('sim_item_rec', 0.06080892608089305)]

# Ranking part

### Обучаем модель 2-ого уровня на выбранных кандидатах

- Обучаем на data_train_ranking
- Обучаем *только* на выбранных кандидатах
- Я *для примера* сгенерирую топ-50 кадидиатов через get_own_recommendations
- (!) Если юзер купил < 50 товаров, то get_own_recommendations дополнит рекоммендации топ-популярными

In [45]:
# 3 временных интервала
# -- давние покупки -- | -- 6 недель -- | -- 3 недель -- 

## Подготовка данных для трейна

In [46]:
N_PREDICT = 50

In [47]:
# взяли пользователей из трейна для ранжирования
df_match_candidates = pd.DataFrame(data_train_ranker[USER_COL].unique())
df_match_candidates.columns = [USER_COL]

In [48]:
# собираем кандитатов с первого этапа (matcher)
df_match_candidates['candidates'] = df_match_candidates[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))

In [49]:
df_match_candidates.head(2)

user_id                                         candidates
0     2070  [1105426, 1097350, 879194, 948640, 928263, 944...
1     2021  [950935, 1119454, 835578, 863762, 1019142, 102...

In [50]:
df_items = df_match_candidates.apply(lambda x: pd.Series(x['candidates']), axis=1).stack().reset_index(level=1, drop=True)
df_items.name = 'item_id'

In [51]:
df_match_candidates = df_match_candidates.drop('candidates', axis=1).join(df_items)

In [52]:
df_match_candidates.head(4)

user_id  item_id
0     2070  1105426
0     2070  1097350
0     2070   879194
0     2070   948640

### Check warm start

In [53]:
print_stats_data(df_match_candidates, 'match_candidates')

match_candidates
Shape: (107550, 2) Users: 2151 Items: 4574


### Создаем трейн сет для ранжирования с учетом кандидатов с этапа 1 

In [71]:
df_ranker_train = data_train_ranker[[USER_COL, ITEM_COL]].copy()
df_ranker_train['target'] = 1  # тут только покупки 

In [72]:
df_ranker_train.head()

user_id  item_id  target
2104867     2070  1019940       1
2107468     2021   840361       1
2107469     2021   856060       1
2107470     2021   869344       1
2107471     2021   896862       1

#### Не хватает нулей в датасете, поэтому добавляем наших кандитатов в качество нулей

In [73]:
df_ranker_train = df_match_candidates.merge(df_ranker_train, on=[USER_COL, ITEM_COL], how='left')

# чистим дубликаты
df_ranker_train = df_ranker_train.drop_duplicates(subset=[USER_COL, ITEM_COL])

df_ranker_train['target'].fillna(0, inplace= True)

In [74]:
df_ranker_train.target.value_counts()

0.0    99177
1.0     7795
Name: target, dtype: int64

In [75]:
df_ranker_train.head(2)

user_id  item_id  target
0     2070  1105426     0.0
1     2070  1097350     0.0

(!) На каждого юзера 50 item_id-кандидатов

In [76]:
df_ranker_train['target'].mean()

0.07286953595333358

- Пока для простоты обучения выберем LightGBM c loss = binary. Это классическая бинарная классификация
- Это пример *без* генерации фич

## Подготавливаем фичи для обучения модели

In [77]:
item_features.head(2)

item_id  manufacturer    department     brand            commodity_desc  \
0    25671             2       GROCERY  National                  FRZN ICE   
1    26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION

In [78]:
user_features.head(2)

age_desc marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   

  household_size_desc kid_category_desc  user_id  
0                   2      None/Unknown        1  
1                   2      None/Unknown        7

In [79]:
df_ranker_train = df_ranker_train.merge(item_features, on='item_id', how='left')
df_ranker_train = df_ranker_train.merge(user_features, on='user_id', how='left')

df_ranker_train.head(2)

user_id  item_id  target  manufacturer department     brand commodity_desc  \
0     2070  1105426     0.0            69       DELI   Private     SANDWICHES   
1     2070  1097350     0.0          2468    GROCERY  National  DOMESTIC WINE   

    sub_commodity_desc curr_size_of_product age_desc marital_status_code  \
0  SANDWICHES - (COLD)                         45-54                   U   
1     VALUE GLASS WINE                4 LTR    45-54                   U   

  income_desc homeowner_desc hh_comp_desc household_size_desc  \
0      50-74K        Unknown      Unknown                   1   
1      50-74K        Unknown      Unknown                   1   

  kid_category_desc  
0      None/Unknown  
1      None/Unknown

In [81]:
data_train_ranker.head()

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
2104867     2070  40618492260  594  1019940         1         1.00       311   
2107468     2021  40618753059  594   840361         1         0.99       443   
2107469     2021  40618753059  594   856060         1         1.77       443   
2107470     2021  40618753059  594   869344         1         1.67       443   
2107471     2021  40618753059  594   896862         2         5.00       443   

         retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
2104867        -0.29          40       86          0.0                0.0  
2107468         0.00         101       86          0.0                0.0  
2107469        -0.09         101       86          0.0                0.0  
2107470        -0.22         101       86          0.0                0.0  
2107471        -2.98         101       86          0.0                0.0

In [82]:
df_matcher = pd.concat([data_train_matcher, data_val_matcher])

In [83]:
df_matcher.head()

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
7      2375  26984851516    1  1085983         1         2.99       364   
11     1364  26984896261    1   999999         1         2.19     31742   
12     1364  26984896261    1   999999         1         2.99     31742   
13     1364  26984896261    1   999999         1         3.09     31742   
14     1364  26984896261    1   937406         1         2.50     31742   

    retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  price  
7         -0.40        1642        1          0.0                0.0   2.99  
11         0.00        1520        1          0.0                0.0   2.19  
12        -0.40        1520        1          0.0                0.0   2.99  
13         0.00        1520        1          0.0                0.0   3.09  
14        -0.99        1520        1          0.0                0.0   2.50

### Добавьте минимум по 2 фичи для юзера, товара и пары юзер-товар

In [84]:
#общая стоимость продаж пользователю
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=USER_COL).agg('sales_value').
                                        sum().rename('total_user_sales'), how='left',on=USER_COL)

# частота покупок user
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=USER_COL).agg(USER_COL).count()
                                        .rename('user_freq'), how='left',on=USER_COL)

#итоговая стоимость продажи товара item
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=ITEM_COL).agg('sales_value').
                                        sum().rename('total_cost_item'), how='left',on=ITEM_COL)

#итоговое значение количества item
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=ITEM_COL).agg('quantity').
                                        sum().rename('total_num_item'), how='left',on=ITEM_COL)
#
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=[USER_COL,ITEM_COL]).agg('quantity').
                                        sum().rename('quantity_user_item'), how='left',on=[USER_COL,ITEM_COL])

#
df_ranker_train = df_ranker_train.merge(df_matcher.groupby(by=[USER_COL,ITEM_COL]).agg('sales_value').
                                        sum().rename('sales_value_user_item'), how='left',on=[USER_COL,ITEM_COL])

In [87]:
df_ranker_train.head()

user_id  item_id  target  manufacturer department     brand  \
0     2070  1105426     0.0            69       DELI   Private   
1     2070  1097350     0.0          2468    GROCERY  National   
2     2070   879194     0.0            69    DRUG GM   Private   
3     2070   948640     0.0          1213    DRUG GM  National   
4     2070   928263     0.0            69    DRUG GM   Private   

          commodity_desc   sub_commodity_desc curr_size_of_product age_desc  \
0             SANDWICHES  SANDWICHES - (COLD)                         45-54   
1          DOMESTIC WINE     VALUE GLASS WINE                4 LTR    45-54   
2  DIAPERS & DISPOSABLES         BABY DIAPERS                14 CT    45-54   
3  ORAL HYGIENE PRODUCTS    WHITENING SYSTEMS                 3 OZ    45-54   
4  DIAPERS & DISPOSABLES         BABY DIAPERS                13 CT    45-54   

   ... homeowner_desc hh_comp_desc household_size_desc kid_category_desc  \
0  ...        Unknown      Unknown                   1      None/Unknown   
1  ...        Unknown      Unknown                   1      None/Unknown   
2  ...        Unknown      Unknown                   1      None/Unknown   
3  ...        Unknown      Unknown                   1      None/Unknown   
4  ...        Unknown      Unknown                   1      None/Unknown   

  total_user_sales user_freq  total_cost_item  total_num_item  \
0          3623.85       796           434.91             109   
1          3623.85       796           619.46              54   
2          3623.85       796           390.81              54   
3          3623.85       796           323.21              49   
4          3623.85       796           451.10              59   

   quantity_user_item  sales_value_user_item  
0                 7.0                  27.93  
1                 NaN                    NaN  
2                 3.0                  21.57  
3                 NaN                    NaN  
4                 3.0                  21.57  

[5 rows x 22 columns]

In [90]:
df_ranker_train[df_ranker_train["item_id"]==1097350].head()

user_id  item_id  target  manufacturer department     brand  \
1        2070  1097350     0.0          2468    GROCERY  National   
1506     1174  1097350     0.0          2468    GROCERY  National   
2860     2498  1097350     0.0          2468    GROCERY  National   
3832      110  1097350     0.0          2468    GROCERY  National   
5061      618  1097350     0.0          2468    GROCERY  National   

     commodity_desc sub_commodity_desc curr_size_of_product age_desc  ...  \
1     DOMESTIC WINE   VALUE GLASS WINE                4 LTR    45-54  ...   
1506  DOMESTIC WINE   VALUE GLASS WINE                4 LTR    25-34  ...   
2860  DOMESTIC WINE   VALUE GLASS WINE                4 LTR    25-34  ...   
3832  DOMESTIC WINE   VALUE GLASS WINE                4 LTR    25-34  ...   
5061  DOMESTIC WINE   VALUE GLASS WINE                4 LTR      NaN  ...   

     homeowner_desc      hh_comp_desc household_size_desc kid_category_desc  \
1           Unknown           Unknown                   1      None/Unknown   
1506        Unknown     2 Adults Kids                  5+                3+   
2860      Homeowner  2 Adults No Kids                   2      None/Unknown   
3832      Homeowner     Single Female                   1      None/Unknown   
5061            NaN               NaN                 NaN               NaN   

     total_user_sales user_freq  total_cost_item  total_num_item  \
1             3623.85       796           619.46              54   
1506           892.73       233           619.46              54   
2860          1834.89       463           619.46              54   
3832          1609.24       414           619.46              54   
5061          3512.18       794           619.46              54   

      quantity_user_item  sales_value_user_item  
1                    NaN                    NaN  
1506                 NaN                    NaN  
2860                 NaN                    NaN  
3832                 NaN                    NaN  
5061                 NaN                    NaN  

[5 rows x 22 columns]

**Фичи user_id:**
    - Средний чек
    - Средняя сумма покупки 1 товара в каждой категории
    - Кол-во покупок в каждой категории
    - Частотность покупок раз/месяц
    - Долю покупок в выходные
    - Долю покупок утром/днем/вечером

**Фичи item_id**:
    - Кол-во покупок в неделю
    - Среднее кол-во покупок 1 товара в категории в неделю
    - (Кол-во покупок в неделю) / (Среднее ол-во покупок 1 товара в категории в неделю)
    - Цена (Можно посчитать из retil_train.csv)
    - Цена / Средняя цена товара в категории
    
**Фичи пары user_id - item_id**
    - (Средняя сумма покупки 1 товара в каждой категории (берем категорию item_id)) - (Цена item_id)
    - (Кол-во покупок юзером конкретной категории в неделю) - (Среднее кол-во покупок всеми юзерами конкретной категории в неделю)
    - (Кол-во покупок юзером конкретной категории в неделю) / (Среднее кол-во покупок всеми юзерами конкретной категории в неделю)

In [92]:
df_ranker_train.head()

user_id  item_id  target  manufacturer department     brand  \
0     2070  1105426     0.0            69       DELI   Private   
1     2070  1097350     0.0          2468    GROCERY  National   
2     2070   879194     0.0            69    DRUG GM   Private   
3     2070   948640     0.0          1213    DRUG GM  National   
4     2070   928263     0.0            69    DRUG GM   Private   

          commodity_desc   sub_commodity_desc curr_size_of_product age_desc  \
0             SANDWICHES  SANDWICHES - (COLD)                         45-54   
1          DOMESTIC WINE     VALUE GLASS WINE                4 LTR    45-54   
2  DIAPERS & DISPOSABLES         BABY DIAPERS                14 CT    45-54   
3  ORAL HYGIENE PRODUCTS    WHITENING SYSTEMS                 3 OZ    45-54   
4  DIAPERS & DISPOSABLES         BABY DIAPERS                13 CT    45-54   

   ... homeowner_desc hh_comp_desc household_size_desc kid_category_desc  \
0  ...        Unknown      Unknown                   1      None/Unknown   
1  ...        Unknown      Unknown                   1      None/Unknown   
2  ...        Unknown      Unknown                   1      None/Unknown   
3  ...        Unknown      Unknown                   1      None/Unknown   
4  ...        Unknown      Unknown                   1      None/Unknown   

  total_user_sales user_freq  total_cost_item  total_num_item  \
0          3623.85       796           434.91             109   
1          3623.85       796           619.46              54   
2          3623.85       796           390.81              54   
3          3623.85       796           323.21              49   
4          3623.85       796           451.10              59   

   quantity_user_item  sales_value_user_item  
0                 7.0                  27.93  
1                 NaN                    NaN  
2                 3.0                  21.57  
3                 NaN                    NaN  
4                 3.0                  21.57  

[5 rows x 22 columns]

In [93]:
X_train = df_ranker_train.drop('target', axis=1)
y_train = df_ranker_train[['target']]

In [94]:
cat_feats = X_train.columns[2:].tolist()
X_train[cat_feats] = X_train[cat_feats].astype('category')

cat_feats

['manufacturer',
 'department',
 'brand',
 'commodity_desc',
 'sub_commodity_desc',
 'curr_size_of_product',
 'age_desc',
 'marital_status_code',
 'income_desc',
 'homeowner_desc',
 'hh_comp_desc',
 'household_size_desc',
 'kid_category_desc',
 'total_user_sales',
 'user_freq',
 'total_cost_item',
 'total_num_item',
 'quantity_user_item',
 'sales_value_user_item']

## Обучение модели ранжирования

In [ ]:
# делали на вебинаре, кусок с катбустом

# from catboost import CatBoost, Pool

# df_bin_feat = pd.get_dummies(X_train)

# model = CatBoost(params ={'objective': "Logloss",
#                           "iterations":1000,
#                           "max_depth":8,
#                           "task_type":"GPU",
#                           "devices":"0:1:2:3"
#             })
# model.fit(df_bin_juice, y_train, silent=True)

# train_preds = model.predict(df_bin_feat,prediction_type="Probability")

In [95]:
lgb = LGBMClassifier(objective='binary',
                     max_depth=8,
                     n_estimators=300,
                     learning_rate=0.05,
                     categorical_column=cat_feats,
                     n_jobs=-1,
#                      verbose=0
                   )

lgb.fit(X_train, y_train)

train_preds = lgb.predict_proba(X_train)

D:\Learning\LibPythonForDS\anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
D:\Learning\LibPythonForDS\anaconda3\lib\site-packages\lightgbm\basic.py:1245: UserWarning: categorical_column in param dict is overridden.
  _log_warning('{} in param dict is overridden.'.format(cat_alias))


In [96]:
df_ranker_predict = df_ranker_train.copy()

In [97]:
df_ranker_predict['proba_item_purchase'] = train_preds[:,1]

In [98]:
df_ranker_predict['proba_item_purchase'][:10]

0    0.149764
1    0.005335
2    0.057791
3    0.000593
4    0.307280
5    0.086047
6    0.004841
7    0.000857
8    0.008639
9    0.549572
Name: proba_item_purchase, dtype: float64

## Подведем итоги

    Мы обучили модель ранжирования на покупках из сета data_train_ranker и на кандитатах от own_recommendations, что является тренировочным сетом, и теперь наша задача предсказать и оценить именно на тестовом сете.

# Evaluation on test dataset

In [99]:
result_eval_ranker = data_val_ranker.groupby(USER_COL)[ITEM_COL].unique().reset_index()
result_eval_ranker.columns=[USER_COL, ACTUAL_COL]
result_eval_ranker.head(2)

user_id                                             actual
0        1  [821867, 834484, 856942, 865456, 889248, 90795...
1        3  [835476, 851057, 872021, 878302, 879948, 90963...

## Eval matching on test dataset

In [100]:
%%time
result_eval_ranker['own_rec'] = result_eval_ranker[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))

Wall time: 7.56 s


In [101]:
# померяем precision только модели матчинга, чтобы понимать влияение ранжирования на метрики

sorted(calc_precision(result_eval_ranker, TOPK_PRECISION), key=lambda x: x[1], reverse=True)

[('own_rec', 0.1444117647058813)]

## Eval re-ranked matched result on test dataset
    Вспомним df_match_candidates сет, который был получен own_recommendations на юзерах, набор пользователей мы фиксировали и он одинаков, значи и прогноз одинаков, поэтому мы можем использовать этот датафрейм для переранжирования.
    

In [102]:
def rerank(user_id):
    return df_ranker_predict[df_ranker_predict[USER_COL]==user_id].sort_values('proba_item_purchase', ascending=False).head(5).item_id.tolist()

In [103]:
result_eval_ranker['reranked_own_rec'] = result_eval_ranker[USER_COL].apply(lambda user_id: rerank(user_id))

In [104]:
print(*sorted(calc_precision(result_eval_ranker, TOPK_PRECISION), key=lambda x: x[1], reverse=True), sep='\n')

('reranked_own_rec', 0.23436031331592475)
('own_rec', 0.1444117647058813)


C:\Users\Al\RecSystems\Lesson6\metrics.py:20: RuntimeWarning: invalid value encountered in long_scalars
  return flags.sum() / len(recommended_list)


#### Вырос ли precision@5 при использовании двухуровневой модели?

вырос

('own_rec', 0.1444117647058813)

'reranked_own_rec', 0.23436031331592475)


Берем топ-k предсказаний, ранжированных по вероятности, для каждого юзера

# Домашнее задание

**Задание 1.**

A) Попробуйте различные варианты генерации кандидатов. Какие из них дают наибольший recall@k ?
- Пока пробуем отобрать 50 кандидатов (k=50)
- Качество измеряем на data_val_matcher: следующие 6 недель после трейна

Дают ли own recommendtions + top-popular лучший recall?  

B)* Как зависит recall@k от k? Постройте для одной схемы генерации кандидатов эту зависимость для k = {20, 50, 100, 200, 500}  
C)* Исходя из прошлого вопроса, как вы думаете, какое значение k является наиболее разумным?


**Задание 2.**

Обучите модель 2-ого уровня, при этом:

- Добавьте минимум по 2 фичи для юзера, товара и пары юзер-товар

- Измерьте отдельно precision@5 модели 1-ого уровня и двухуровневой модели на data_val_ranker

- Вырос ли precision@5 при использовании двухуровневой модели?